File: rag-engine/documents/validation.py
"""Validate document-builder input against a template."""

from __future__ import annotations

import re
from datetime import date, datetime
from typing import Any, Mapping

from .conditions import clause_is_active, has_value
from .constants import INDIA_REGIONS
from .types import FieldSpec, TemplateSpec

EMAIL_RE = re.compile(r"^[^@\s]+@[^@\s]+\.[^@\s]+$")
ISO_DATE_RE = re.compile(r"^(\d{4})-(\d{2})-(\d{2})$")


class FieldIssue(dict):
    """JSON-serialisable validation issue."""


def _string(value: Any) -> str:
    return value.strip() if isinstance(value, str) else ""


def validate_field(field: FieldSpec, values: Mapping[str, Any]) -> list[dict[str, str]]:
    field_id = field["id"]
    raw = values.get(field_id)
    requirement = field["requirement"]
    present = has_value(values, field_id)
    issues: list[dict[str, str]] = []

    if not present:
        if requirement == "required":
            issues.append({"field_id": field_id, "level": "error",
                           "message": f"{field['label']} is required before a draft can be generated."})
        elif requirement == "recommended":
            issues.append({"field_id": field_id, "level": "recommended",
                           "message": f"{field['label']} is recommended. The draft will mark it for completion if this clause is used."})
        return issues

    field_type = field["type"]
    if field_type in {"text", "textarea", "tel"} and not _string(raw):
        issues.append({"field_id": field_id, "level": "error", "message": f"{field['label']} cannot be empty."})
    if field_type == "email" and not EMAIL_RE.match(_string(raw)):
        issues.append({"field_id": field_id, "level": "error", "message": f"{field['label']} must be a valid email address."})
    if field_type == "date":
        match = ISO_DATE_RE.match(_string(raw))
        valid = False
        if match:
            try:
                date(int(match.group(1)), int(match.group(2)), int(match.group(3)))
                valid = True
            except ValueError:
                valid = False
        if not valid:
            issues.append({"field_id": field_id, "level": "error", "message": f"{field['label']} must be a valid date."})
    if field_type in {"number", "currency"}:
        try:
            number = float(raw)
        except (TypeError, ValueError):
            issues.append({"field_id": field_id, "level": "error", "message": f"{field['label']} must be a number."})
        else:
            if "min" in field and number < field["min"]:
                issues.append({"field_id": field_id, "level": "error",
                               "message": f"{field['label']} must be at least {field['min']}."})
            if "max" in field and number > field["max"]:
                issues.append({"field_id": field_id, "level": "error",
                               "message": f"{field['label']} must be at most {field['max']}."})
    if field_type == "select" and field.get("options"):
        allowed = {option["value"] for option in field["options"]}
        if str(raw) not in allowed:
            issues.append({"field_id": field_id, "level": "error",
                           "message": f"{field['label']} must be one of the listed options."})
    if field_id == "jurisdiction_country" and _string(raw) != "IN":
        issues.append({"field_id": field_id, "level": "error",
                       "message": "This generator currently supports India only."})
    if field_id == "jurisdiction_region" and _string(raw) not in INDIA_REGIONS:
        issues.append({"field_id": field_id, "level": "error",
                       "message": "Select a State or Union Territory in India."})
    if field.get("pattern") and _string(raw) and not re.match(field["pattern"], _string(raw)):
        issues.append({"field_id": field_id, "level": "error", "message": f"{field['label']} is not in the expected format."})
    return issues


def validate_values(template: TemplateSpec, values: Mapping[str, Any]) -> dict[str, Any]:
    issues: list[dict[str, str]] = []
    for field in template["fields"]:
        issues.extend(validate_field(field, values))
    start = _string(values.get("effective_date"))
    end = _string(values.get("end_date"))
    if start and end and ISO_DATE_RE.match(start) and ISO_DATE_RE.match(end):
        start_date = datetime.strptime(start, "%Y-%m-%d").date()
        end_date = datetime.strptime(end, "%Y-%m-%d").date()
        if end_date < start_date:
            issues.append({"field_id": "end_date", "level": "error",
                           "message": "The end date cannot be earlier than the start date."})
    blocking = [issue for issue in issues if issue["level"] == "error"]
    return {
        "ok": not blocking,
        "blocking": blocking,
        "issues": issues,
        "can_generate": not blocking,
    }


def active_clauses(template: TemplateSpec, values: Mapping[str, Any]) -> list[dict[str, Any]]:
    selected = []
    for clause in template["clauses"]:
        if clause_is_active(clause.get("condition"), values):
            selected.append(clause)
    return selected


def placeholder_for(field: FieldSpec) -> str:
    return f"[{field['placeholder_label']} REQUIRED]"


CRITICAL_PLACEHOLDER_RE = re.compile(r"\[[A-Z][A-Z0-9 /._-]{2,} REQUIRED\]")


def unresolved_critical_placeholders(template: TemplateSpec, values: Mapping[str, Any], sections: list[dict[str, Any]]) -> list[str]:
    fields = {field["id"]: field for field in template["fields"]}
    required_placeholders = {
        placeholder_for(field) for field in template["fields"] if field["requirement"] == "required"
    }
    found: list[str] = []
    for section in sections:
        body = section.get("body") or ""
        for placeholder in required_placeholders:
            if placeholder in body:
                field_id = next(fid for fid, field in fields.items() if placeholder_for(field) == placeholder)
                if has_value(values, field_id):
                    found.append(placeholder)
                elif fields[field_id]["requirement"] == "required":
                    found.append(placeholder)
    return list(dict.fromkeys(found))

File: web/lib/documents/validation.ts
import { clauseIsActive, hasValue } from "./conditions";
import { INDIA_REGIONS } from "./constants";
import type { DocumentValues, FieldIssue, FieldSpec, TemplateSpec } from "./types";

const EMAIL_RE = /^[^@\s]+@[^@\s]+\.[^@\s]+$/;
const ISO_DATE_RE = /^(\d{4})-(\d{2})-(\d{2})$/;

function asString(value: unknown): string {
  return typeof value === "string" ? value.trim() : "";
}

export function validateField(field: FieldSpec, values: DocumentValues): FieldIssue[] {
  const present = hasValue(values, field.id);
  const raw = values[field.id];
  if (!present) {
    if (field.requirement === "required") {
      return [{ fieldId: field.id, level: "error", message: `${field.label} is required before a draft can be generated.` }];
    }
    if (field.requirement === "recommended") {
      return [{ fieldId: field.id, level: "recommended", message: `${field.label} is recommended.` }];
    }
    return [];
  }
  const issues: FieldIssue[] = [];
  if ((field.type === "text" || field.type === "textarea" || field.type === "tel") && !asString(raw)) {
    issues.push({ fieldId: field.id, level: "error", message: `${field.label} cannot be empty.` });
  }
  if (field.type === "email" && !EMAIL_RE.test(asString(raw))) {
    issues.push({ fieldId: field.id, level: "error", message: `${field.label} must be a valid email address.` });
  }
  if (field.type === "date") {
    const match = asString(raw).match(ISO_DATE_RE);
    const valid = Boolean(match && !Number.isNaN(Date.parse(`${match[1]}-${match[2]}-${match[3]}`)));
    if (!valid) issues.push({ fieldId: field.id, level: "error", message: `${field.label} must be a valid date.` });
  }
  if (field.type === "number" || field.type === "currency") {
    const number = typeof raw === "number" ? raw : Number(raw);
    if (!Number.isFinite(number)) {
      issues.push({ fieldId: field.id, level: "error", message: `${field.label} must be a number.` });
    } else {
      if (field.min !== undefined && number < field.min) {
        issues.push({ fieldId: field.id, level: "error", message: `${field.label} must be at least ${field.min}.` });
      }
      if (field.max !== undefined && number > field.max) {
        issues.push({ fieldId: field.id, level: "error", message: `${field.label} must be at most ${field.max}.` });
      }
    }
  }
  if (field.type === "select" && field.options && !field.options.some((option) => option.value === String(raw))) {
    issues.push({ fieldId: field.id, level: "error", message: `${field.label} must be one of the listed options.` });
  }
  if (field.id === "jurisdiction_country" && asString(raw) !== "IN") {
    issues.push({ fieldId: field.id, level: "error", message: "This generator currently supports India only." });
  }
  if (field.id === "jurisdiction_region" && !INDIA_REGIONS.includes(asString(raw) as (typeof INDIA_REGIONS)[number])) {
    issues.push({ fieldId: field.id, level: "error", message: "Select a State or Union Territory in India." });
  }
  return issues;
}

export function validateValues(template: TemplateSpec, values: DocumentValues) {
  const issues = template.fields.flatMap((item) => validateField(item, values));
  const start = asString(values.effective_date);
  const end = asString(values.end_date);
  if (start && end && ISO_DATE_RE.test(start) && ISO_DATE_RE.test(end) && end < start) {
    issues.push({ fieldId: "end_date", level: "error", message: "The end date cannot be earlier than the start date." });
  }
  const blocking = issues.filter((issue) => issue.level === "error");
  return { ok: blocking.length === 0, blocking, issues, canGenerate: blocking.length === 0 };
}

export function activeClauses(template: TemplateSpec, values: DocumentValues) {
  return template.clauses.filter((clause) => clauseIsActive(clause.condition, values));
}

export function placeholderFor(field: FieldSpec): string {
  return `[${field.placeholderLabel} REQUIRED]`;
}

export function stepHasErrors(template: TemplateSpec, stepId: string, values: DocumentValues): boolean {
  const step = template.steps.find((item) => item.id === stepId);
  if (!step) return false;
  return step.fieldIds.some((fieldId) => {
    const field = template.fields.find((item) => item.id === fieldId);
    return field ? validateField(field, values).some((issue) => issue.level === "error") : false;
  });
}
